File: sdk/python/rover_client.py
"""Small standard-library client for Rover's CLI (v0.x API).
No shell, automatic retries, provider credentials, or acceptance inference.
A CLI submission may continue after a client timeout; reconcile before retrying.
"""
from __future__ import annotations
from dataclasses import dataclass
import json
import os
from pathlib import Path
import subprocess
import tempfile
from typing import Any, Sequence

class RoverError(RuntimeError):
    """Protocol, invocation or bounded-output failure."""

@dataclass(frozen=True)
class Result:
    exit_code: int
    data: Any

    @property
    def decision(self) -> str | None:
        return self.data.get("decision") if isinstance(self.data, dict) else None

class Rover:
    def __init__(self, binary: str | os.PathLike[str], state: str | os.PathLike[str], *, timeout: float = 60, max_output: int = 16 << 20):
        self.binary = str(Path(binary).resolve())
        self.state = str(Path(state).resolve())
        if timeout <= 0 or max_output < 1024:
            raise ValueError("positive timeout and output bound required")
        self.timeout = timeout
        self.max_output = max_output

    def call(self, args: Sequence[str]) -> Result:
        if isinstance(args, str) or any(not isinstance(x, str) for x in args):
            raise TypeError("args must be an argv sequence, not a shell string")
        argv = [self.binary, "--state", self.state, *args, "--json"]
        # Avoid accumulating unbounded pipe output in memory. Rover's own output
        # limits remain relevant; operating-system disk quotas are still needed.
        with tempfile.TemporaryFile() as stdout, tempfile.TemporaryFile() as stderr:
            try:
                proc = subprocess.run(argv, stdin=subprocess.DEVNULL, stdout=stdout, stderr=stderr, timeout=self.timeout, check=False)
            except subprocess.TimeoutExpired as exc:
                raise RoverError("Client timed out. An admitted task may still be running; inspect state before retrying.") from exc
            stdout.seek(0); raw = stdout.read(self.max_output + 1)
            stderr.seek(0); err = stderr.read(65536).decode("utf-8", "replace")
            if len(raw) > self.max_output:
                raise RoverError("Rover output exceeded configured client limit")
            try:
                data = json.loads(raw)
            except (ValueError, UnicodeError) as exc:
                raise RoverError(f"Rover did not return JSON (exit {proc.returncode}): {err}") from exc
            return Result(proc.returncode, data)

    def inspect(self, repo: str, *, base: str = "HEAD", worktree: bool = False) -> Result:
        return self.call(["inspect", "--repo", repo, "--base", base] + (["--worktree"] if worktree else []))

    def verify(self, repo: str, *, base: str = "HEAD", allow_local: bool = False, worktree: bool = False) -> Result:
        return self.call(["verify", "--repo", repo, "--base", base] + (["--allow-local"] if allow_local else []) + (["--worktree"] if worktree else []))

    def run_task(self, task_file: str, *, allow_local: bool = False, key: str | None = None) -> Result:
        argv = ["task", "run", "--file", task_file]
        if allow_local: argv.append("--allow-local")
        if key is not None: argv += ["--key", key]
        return self.call(argv)

    def status(self, task_id: str | None = None) -> Result:
        return self.call(["status"] + (["--id", task_id] if task_id else []))

    def report(self, investigation_id: str) -> Result:
        return self.call(["report", "--id", investigation_id])

File: sdk/typescript/src/rover_client.ts
/**
 * Small Node client for Rover's CLI (v0.x API).
 * No shell, automatic retries, provider credentials, or acceptance inference.
 * A CLI submission may continue after a client timeout; reconcile before retrying.
 */
import { spawnSync } from "node:child_process";
import * as path from "node:path";

export class RoverError extends Error {
  override name = "RoverError";
}

export interface Result {
  exitCode: number;
  data: unknown;
  /** convenience when data is {decision: string} */
  get decision(): string | null;
}

class ResultImpl implements Result {
  exitCode: number;
  data: unknown;
  constructor(exitCode: number, data: unknown) {
    this.exitCode = exitCode;
    this.data = data;
  }
  get decision(): string | null {
    if (dataIsRecord(this.data) && typeof this.data["decision"] === "string") return this.data["decision"] as string;
    return null;
  }
}

function dataIsRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null;
}

export class Rover {
  readonly binary: string;
  readonly state: string;
  readonly timeoutMs: number;
  readonly maxOutput: number;

  constructor(binary: string, state: string, opts: { timeoutMs?: number; maxOutput?: number } = {}) {
    this.binary = path.resolve(binary);
    this.state = path.resolve(state);
    this.timeoutMs = opts.timeoutMs ?? 60_000;
    this.maxOutput = opts.maxOutput ?? 16 * 1024 * 1024;
    if (this.timeoutMs <= 0 || this.maxOutput < 1024) throw new RangeError("positive timeout and output bound required");
  }

  call(args: string[]): Result {
    if (typeof args === "string" || !Array.isArray(args) || args.some((x) => typeof x !== "string")) {
      throw new TypeError("args must be an argv string[] , not a shell string");
    }
    const argv = [this.binary, "--state", this.state, ...args, "--json"];
    // spawnSync avoids shell; timeout is OS-level. Rover's own output limits remain relevant.
    const r = spawnSync(argv[0], argv.slice(1), {
      input: undefined,
      timeout: this.timeoutMs,
      maxBuffer: this.maxOutput + 1024,
      encoding: "buffer",
    } as const);

    if (r.error) {
      const e = r.error as NodeJS.ErrnoException;
      if (e.code === "ETIMEDOUT" || r.signal === "SIGTERM") {
        throw new RoverError("Client timed out. An admitted task may still be running; inspect state before retrying.");
      }
      throw new RoverError(String(e.message));
    }

    const stdout = r.stdout as Buffer | null;
    const stderr = r.stderr as Buffer | null;
    const raw = stdout ?? Buffer.alloc(0);
    const errText = (stderr ?? Buffer.alloc(0)).toString("utf8", 0, 65536).replace(/\0/g, "");

    if (raw.length > this.maxOutput) throw new RoverError("Rover output exceeded configured client limit");

    let data: unknown;
    try {
      data = JSON.parse(raw.toString("utf8"));
    } catch (e) {
      throw new RoverError(`Rover did not return JSON (exit ${r.status ?? -1}): ${errText.slice(0, 400)}`);
    }
    return new ResultImpl(r.status ?? 0, data);
  }

  inspect(repo: string, opts: { base?: string; worktree?: boolean } = {}): Result {
    const a = ["inspect", "--repo", repo, "--base", opts.base ?? "HEAD"];
    if (opts.worktree) a.push("--worktree");
    return this.call(a);
  }

  verify(repo: string, opts: { base?: string; allowLocal?: boolean; worktree?: boolean } = {}): Result {
    const a = ["verify", "--repo", repo, "--base", opts.base ?? "HEAD"];
    if (opts.allowLocal) a.push("--allow-local");
    if (opts.worktree) a.push("--worktree");
    return this.call(a);
  }

  runTask(taskFile: string, opts: { allowLocal?: boolean; key?: string } = {}): Result {
    const a = ["task", "run", "--file", taskFile];
    if (opts.allowLocal) a.push("--allow-local");
    if (opts.key) a.push("--key", opts.key);
    return this.call(a);
  }

  status(taskId?: string): Result {
    return this.call(taskId ? ["status", "--id", taskId] : ["status"]);
  }

  report(investigationId: string): Result {
    return this.call(["report", "--id", investigationId]);
  }
}
